### DISCUSSION TITLE
Newbie Question about the AspNetCore example

### DISCUSSION THREAD

**Discussion by GoOnYahGoodThing** [quartznet/quartznet#1569] [Q&A]
Hi,
I downloaded and built the sources using VS2022 Community edition.
Then I set the Quartz.Examples.AspNetCore project as the startup project and ran it.

Now I see a home page that doesn't appear to do much ??
I see a list of Scheduled Triggers... and then the Currently Executing Jobs is empty. And it stays empty no matter how long I wait.
The startup.cs would appear to be adding jobs and triggers to the scheduler.. but they don't appear to be being triggered?

I set breakpoints in the ExampleJob and the SlowJob and the breakpoints are never hit..also the log output to the console never shows any output from the jobs.

So I'm wondering if I am doing something wrong? Or if the project will run in Community Edition?

Sorry if this is a dumb newbie post.. but I am just trying to get started... there's so much in this example project and its not very obvious what its all for.. like the Health Checks UI ? Whats that for ? 

This all looks like a brilliant example.. but I am really missing some background or explanation .. is there a page somewhere I could use to follow this code?
Thanks

**Comment by jafin** [comment]
Bit of a segue, as I'm not answering your actual question just offering an alternative

There are simpler examples in the https://github.com/quartznet/quartznet/tree/main/src/Quartz.Examples  Quartz.Examples project.  These isolate what they are showing without the overhead of mixing in web activities. Perhaps have a look there first if you are some basic quartz functionality demos.

**Reply by lahma** [reply]
@GoOnYahGoodThing There seems to be a bit of a problem in the current web sample application after the v4 merge to main. To test things out you should probably use the 3.x branch which should be quite solid and corresponds to the current NuGet release.

I'll look at the web sample a bit later and fix it.

**Reply by GoOnYahGoodThing** [reply]
Thank you @lahma 
Great to know its not just Me.. I'll try the 3.x branch for now so.. and will be excited to see the 6 version !
Thanks

**Reply by GoOnYahGoodThing** [reply]
Ohh boy.. when I try to build the 3.x code , it complains that framework 4.6.2 isnt installed.

So i downloaded the installer for the developer pack and ran it.. but it said 

".NET Framework 4.6.2 or a later update is already installed on this computer."

Jeez.. one of those days when nothing is going to go right..

**Reply by GoOnYahGoodThing** [reply]
Thank God for Stack overflow.. 
Visual Studio Installer -> Modify -> Individual Components and check the right version
worked.. heh..

**Reply by lahma** [reply]
@GoOnYahGoodThing the ASP.NET Core example should now work also on main branch. Thanks for reporting the problems.

**Comment by GoOnYahGoodThing** [comment]
Thanks @jafin , I should have said that I did actually have a look at those examples and think I understood most of what I needed to there. 
But ultimately I will want to use Quartz in an ASP.NET Core 6 app so I was just hoping to see something working and understand where to start jobs etc. 
Even in the AspNetCore example all the jobs are being scheduled in startup.cs and thats also not really how I want to proceed. 
So what I'm hoping to understand is how to resolve the service using DI and then interrogate it to get a list of jobs and do things like interrupt a selected job and maybe start a job that has been created as durable..  
So the AspNetCore example does seem to show that in the Index.cshtml.cs file but the GetCurrentlyExecutingJobs method keeps returning an empty list... and I wasn't sure if there was a problem there or not.
I would hope to contribute a very simple example of an ASP.NET Core 6 app that starts and stops jobs (when the user clicks a button or two) to the project eventually.
Thanks again for taking the time to respond to my question.

**Comment by Sicos1977** [comment]
I also sometimes have that problem from Visual Studio. The solution is to install the .NET 4.6.2 framework from the Visual Studio installer so that it knows it is installed.

![image](https://user-images.githubusercontent.com/6692947/161038895-b86fc55a-efae-418f-9f93-cbd09c499020.png)

![image](https://user-images.githubusercontent.com/6692947/161038592-1211061d-2afc-4925-9c20-83b7ee942bd9.png)

**Reply by GoOnYahGoodThing** [reply]
Brilliant! Thanks.
Now the 3.x branch version of Quartz.Examples.AspNetCore project is working great.
![image](https://user-images.githubusercontent.com/74314759/161051528-502ba902-6bf1-48fc-aec2-3d98ba9b87cf.png)

**Comment by GoOnYahGoodThing** [comment]
Ok, so now, Im recovered from the covid.

I am back to look and try to understand this example.
And I am trying to convert it to use the new .net core 6.0 style of minimal hosting model without the use of startup.cs.

This is a great example.. there is so much in it.
I'm sorry but.. there are so many things in it that I find it hard to understand.. so I am trying to cut it down to just one simple job...

This code appears to work.. but I would like to know if there is anything missing.. as the start of the log info says the scheduler is not started.. and then it just starts running the job anyway.. so I'm thinking I must be doing something wrong...

The log output is like this..

```
info: Quartz.Core.SchedulerSignalerImpl[0]
      Initialized Scheduler Signaller of type: Quartz.Core.SchedulerSignalerImpl
info: Quartz.Core.QuartzScheduler[0]
      Quartz Scheduler created
info: Quartz.Core.QuartzScheduler[0]
      JobFactory set to: Quartz.Simpl.MicrosoftDependencyInjectionJobFactory
info: Quartz.Simpl.RAMJobStore[0]
      RAMJobStore initialized.
info: Quartz.Core.QuartzScheduler[0]
      Scheduler meta-data: Quartz Scheduler (v3.4.0.0) 'Quartz ASP.NET Core Sample Scheduler' with instanceId 'Scheduler-Core'
  Scheduler class: 'Quartz.Core.QuartzScheduler' - running locally.
  NOT STARTED.
  Currently in standby mode.
  Number of jobs executed: 0
  Using thread pool 'Quartz.Simpl.DefaultThreadPool' - with 10 threads.
  Using job-store 'Quartz.Simpl.RAMJobStore' - which does not support persistence. and is not clustered.

info: Quartz.Impl.StdSchedulerFactory[0]
      Quartz scheduler 'Quartz ASP.NET Core Sample Scheduler' initialized
info: Quartz.Impl.StdSchedulerFactory[0]
      Quartz scheduler version: 3.4.0.0
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:7109
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5109
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\pjmealy\source\learning\QuartzPOC\QuartzPOC\
info: Quartz.Core.QuartzScheduler[0]
      Scheduler Quartz ASP.NET Core Sample Scheduler_$_Scheduler-Core started.
info: QuartzPOC.ExampleJob[0]
      DEFAULT.Combined Configuration Trigger job executing, triggered by DEFAULT.Combined Configuration Trigger
info: QuartzPOC.ExampleJob[0]
      EXAMPLE JOB STARTING
info: QuartzPOC.ExampleJob[0]
      DEFAULT.Combined Configuration Trigger job executing, triggered by DEFAULT.Combined Configuration Trigger
info: QuartzPOC.ExampleJob[0]
      EXAMPLE JOB STARTING
info: QuartzPOC.ExampleJob[0]
      EXAMPLE JOB ENDING
info: QuartzPOC.ExampleJob[0]
      Example job disposing
```

So its funny cos it starts out says NOT STARTED and Currently in standby mode.
But then goes on to initialize etc.. maybe its right.. but I'd like to be sure.

The code for the .net core 6.0 minimal hosting style Program.cs that I made is like this:

```
using System;
using System.Globalization;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Quartz;
using Quartz.Impl.Calendar;
using Quartz.Impl.Matchers;
using Quartz.Plugin.Interrupt;
using QuartzPOC;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddRazorPages();

// base configuration for DI, read from appSettings.json
builder.Services.Configure<QuartzOptions>(builder.Configuration.GetSection("Quartz"));

// if you are using persistent job store, you might want to alter some options
builder.Services.Configure<QuartzOptions>(options =>
{
    options.Scheduling.IgnoreDuplicates = true; // default: false
    options.Scheduling.OverWriteExistingData = true; // default: true
});

builder.Services.AddQuartz(q =>
{
    // handy when part of cluster or you want to otherwise identify multiple schedulers
    q.SchedulerId = "Scheduler-Core";

    // you can control whether job interruption happens for running jobs when scheduler is shutting down
    q.InterruptJobsOnShutdown = true;

    // when QuartzHostedServiceOptions.WaitForJobsToComplete = true or scheduler.Shutdown(waitForJobsToComplete: true)
    q.InterruptJobsOnShutdownWithWait = true;

    // we can change from the default of 1
    q.MaxBatchSize = 5;

    // we take this from appsettings.json, just show it's possible
    // q.SchedulerName = "Quartz ASP.NET Core Sample Scheduler";

    // this is default configuration if you don't alter it
    q.UseMicrosoftDependencyInjectionJobFactory();

    // these are the defaults
    q.UseSimpleTypeLoader();
    q.UseInMemoryStore();
    q.UseDefaultThreadPool(maxConcurrency: 10);

    // quickest way to create a job with single trigger is to use ScheduleJob
    q.ScheduleJob<ExampleJob>(trigger => trigger
        .WithIdentity("Combined Configuration Trigger")
        .StartAt(DateBuilder.EvenSecondDate(DateTimeOffset.UtcNow.AddSeconds(7)))
        .WithDailyTimeIntervalSchedule(x => x.WithInterval(10, IntervalUnit.Second))
        .WithDescription("SingleJob trigger configured for a job with single call")
    );

    // you can also configure individual jobs and triggers with code
    // this allows you to associated multiple triggers with same job
    // (if you want to have different job data map per trigger for example)
    q.AddJob<ExampleJob>(j => j
        .StoreDurably() // we need to store durably if no trigger is associated
        .WithDescription("my durable job")
    );

    // convert time zones using converter that can handle Windows/Linux differences
    q.UseTimeZoneConverter();
});

// we can use options pattern to support hooking your own configuration with Quartz's
// because we don't use service registration api, we need to manually ensure the job is present in DI
builder.Services.AddTransient<ExampleJob>();

// ASP.NET Core hosting
builder.Services.AddQuartzServer(options =>
{
    // when shutting down we want jobs to complete gracefully
    options.WaitForJobsToComplete = true;
});

// ------------------------------------------------------------
var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();
app.UseStaticFiles();

app.UseRouting();

app.UseAuthorization();

app.MapRazorPages();

app.Run();

```

**Reply by lahma** [reply]
`builder.Services.AddQuartzServer` will add Quartz Hosted Service which will also start the scheduler.

Omit the line and then you need to manually resolve the scheduler and start it.

**Reply by GoOnYahGoodThing** [reply]
Ahh ok.. so that line starts the scheduler so I don't need to do:
```
scheduler.Start();
```

Great.. its starting to fall into place now.. I have a page with a button that can start a job etc.
```
<div class="row">
    
        <form method="post">
            <p>
                <div class="form-group">
                    
                    <input type="submit" value="Run new Job" asp-page-handler="StartJob" class="btn btn-primary" />
                </div>
            </p>
        </form>
 
</div>
```

```
public async Task<IActionResult> OnPostStartJob()
        {
            var scheduler = await schedulerFactory.GetScheduler();


            // define the job and tie it to our HelloJob class
            IJobDetail job = JobBuilder.Create<ExampleJob>()
                .WithIdentity("job1", "group1")
                .Build();

            // Trigger the job to run now, and then repeat every 10 seconds
            ITrigger trigger = TriggerBuilder.Create()
                .WithIdentity("trigger1", "group1")
                .StartNow()
                .Build();
            // Tell quartz to schedule the job using our trigger
            await scheduler.ScheduleJob(job, trigger);

            Triggers = await scheduler.GetTriggerKeys(GroupMatcher<TriggerKey>.AnyGroup());
            await Task.Delay(TimeSpan.FromSeconds(1));
            CurrentlyExecutingJobs = await scheduler.GetCurrentlyExecutingJobs();

            return Page();
        }
```
I just have a 1 second delay in there cos the currently executing jobs doesn't get updated immediately...

Other than that its moving along nicely.
Thanks @lahma !
